File: js/download.js
const https = require("https");
const fs = require("fs");
const path = require("path");

const MODEL_URLS = {
  onnx: "https://huggingface.co/janakh/monocr/resolve/main/monocr.onnx",
  tflite: "https://huggingface.co/janakh/monocr/resolve/main/monocr.tflite",
  charset: "https://huggingface.co/janakh/monocr/resolve/main/charset.txt",
};

async function downloadFile(url, dest) {
  return new Promise((resolve, reject) => {
    const file = fs.createWriteStream(dest);
    https
      .get(url, (response) => {
        if (response.statusCode !== 200) {
          reject(
            new Error(`Failed to download ${url}: ${response.statusCode}`),
          );
          return;
        }
        response.pipe(file);
        file.on("finish", () => {
          file.close(resolve);
        });
      })
      .on("error", (err) => {
        fs.unlink(dest, () => reject(err));
      });
  });
}

async function main() {
  const destDir = process.argv[2] || "model";
  if (!fs.existsSync(destDir)) {
    fs.mkdirSync(destDir, { recursive: true });
  }

  for (const [name, url] of Object.entries(MODEL_URLS)) {
    const fileName = name === "charset" ? "charset.txt" : `monocr.${name}`;
    const dest = path.join(destDir, fileName);
    console.log(`Downloading ${url} to ${dest}...`);
    await downloadFile(url, dest);
  }
  console.log("Done.");
}

if (require.main === module) {
  main().catch(console.error);
}

File: python/monocr_onnx/download.py
import argparse
import requests
import os
from pathlib import Path

MODEL_URLS = {
    "onnx": "https://huggingface.co/janakh/monocr/resolve/main/monocr.onnx",
    "tflite": "https://huggingface.co/janakh/monocr/resolve/main/monocr.tflite",
    "charset": "https://huggingface.co/janakh/monocr/resolve/main/charset.txt",
}

def download_file(url, dest):
    print(f"Downloading {url} to {dest}...")
    response = requests.get(url, stream=True)
    response.raise_for_status()
    
    os.makedirs(os.path.dirname(dest), exist_ok=True)
    with open(dest, "wb") as f:
        for chunk in response.iter_content(chunk_size=8192):
            f.write(chunk)
    print(f"Done.")

def main():
    parser = argparse.ArgumentParser(description="Download MonOCR models")
    parser.add_argument("--dest", type=str, default="model", help="Destination directory")
    args = parser.parse_args()
    
    dest_path = Path(args.dest)
    
    for name, url in MODEL_URLS.items():
        ext = ".txt" if name == "charset" else f".{name}"
        file_dest = dest_path / f"monocr{ext}" if name != "charset" else dest_path / "charset.txt"
        download_file(url, file_dest)

if __name__ == "__main__":
    main()
